src/main/java/org/stellar/sdk/requests/SSEStream.java

```diff
@@ -0,0 +1,180 @@
+package org.stellar.sdk.requests;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.internal.sse.RealEventSource;
+import okhttp3.sse.EventSource;
+import okhttp3.sse.EventSourceListener;
+import org.stellar.sdk.responses.GsonSingleton;
+
+import javax.annotation.Nullable;
+import java.io.Closeable;
+import java.net.SocketException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+public class SSEStream<T extends org.stellar.sdk.responses.Response> implements Closeable {
+  private final OkHttpClient okHttpClient;
+  private final RequestBuilder requestBuilder;
+  private final Class<T> responseClass;
+  private final EventListener<T> listener;
+
+  private final AtomicBoolean isStopped = new AtomicBoolean(false);
+  private final AtomicBoolean serverSideClosed = new AtomicBoolean(true); // make sure we start correctly
+  private final AtomicReference<String> pagingToken = new AtomicReference<String>("0");
+  private final AtomicReference<String> lastEventId = new AtomicReference<String>(null);
```

Comment:
That shouldn't happen but since we're using threads, I'd use it on principle here just in case. This shortcuts the potential for a lot of subtle concurrency bugs.